### ISSUE TITLE
amp tests fail on sixth run

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2670]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2670|
|Type|defect|
|Created|2007-05-21 00:11:23Z|

On the sixth test run of trunk@20346, ```trial -u twisted.test.test_amp``` fails with these errors:

```
===============================================================================
[ERROR]: twisted.test.test_amp.AMPTest.test_helloNoErrorHandling

Traceback (most recent call last):
Failure: OpenSSL.SSL.Error: [('SSL routines', 'SSL3_WRITE_BYTES', 'ssl handshake failure')]
===============================================================================
[ERROR]: twisted.test.test_amp.AMPTest.test_helloNoErrorHandling

Traceback (most recent call last):
Failure: OpenSSL.SSL.Error: [('SSL routines', 'SSL3_WRITE_BYTES', 'ssl handshake failure')]
===============================================================================
[ERROR]: twisted.test.test_amp.AMPTest.test_helloNoErrorHandling

Traceback (most recent call last):
Failure: OpenSSL.SSL.Error: [('SSL routines', 'SSL3_WRITE_BYTES', 'ssl handshake failure')]
===============================================================================
[ERROR]: twisted.test.test_amp.AMPTest.test_helloNoErrorHandling

Traceback (most recent call last):
Failure: OpenSSL.SSL.Error: [('SSL routines', 'SSL3_WRITE_BYTES', 'ssl handshake failure')]
-------------------------------------------------------------------------------
```

Attachments:

* [test_amp_2670.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5c4/5c46489bec06fb161c255fbb3f09e836b23b08dc/4b91f4eeb41999d9988291eeccc62124150b2414.diff) (1230 bytes) - added by therve on 2007-05-22 13:35:19Z - 
* [test_amp_2_2670.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5c4/5c46489bec06fb161c255fbb3f09e836b23b08dc/dec0296b7bea7f8d9606f22d8ab69fa0f0acaabc.diff) (3695 bytes) - added by therve on 2007-05-24 10:22:48Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2670 2670
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1179706283000000 1179706283000000
changetime__1180163282000000 1180163282000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_1"></a>|@radix commented|
|-|-|

On my Feisty compy it's the third run, and it's a different test. but it is deterministic.

Basically OPENSSL is garbage Or maybe PyOPENSSL

```
===============================================================================
[ERROR]: twisted.test.test_amp.AMPTest.test_brokenReturnValue

Traceback (most recent call last):
Failure: OpenSSL.SSL.Error: [('SSL routines', 'SSL3_WRITE_BYTES', 'ssl handshake failure')]
-------------------------------------------------------------------------------
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve removed owner|
|-|-|

It turns out that the test was completely broken, and in fact fails every time. Attached a patch correcting it.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_3"></a>|@radix commented|
|-|-|

This patch isn't applying cleanly to trunk, and unfortunately I've never learned how to resolve patch rejections (man, why doesn't patch have an option to do the thing that svn does on a conflicting merge?)

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_4"></a>|@radix set owner to @therve|
|-|-|

This is of course a great patch, and fixes up some really dumb mistakes.

If you change the gatherResults(L) to gatherResults(L).addErrback(lambda first: first.value.subFailure), then +1.

However, the tearDown here is really really terrible. If you want to fix it up so the testcases use TestCase.addCleanup, that would improve the situation a lot and I would be very happy :-)

Thanks for the patch!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @radix|
|-|-|

I updated the patch. I cleanup tearDown a bit, and added some docstrings.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @therve|
|-|-|

I can't figure out the meaning of the class docstring of SlightlySmartTLS.  Specifically, I can't fit any meaning to "the protocol user server side".

Also, ugg, these tests are terrible. :(

The rest of the diff looks good to me, though.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

(In [20403]) Apply patch, correct doc typo

Refs [#2670](https://github.com/twisted/twisted/issues/2670)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve removed owner|
|-|-|

OK I applied the patch in amp-ssl-test-2670 to ease review :). I also corrected the typo in the docstring of SlightlySmartTLS.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_9"></a>|@radix set owner to @therve|
|-|-|

It doesn't fix the worst of the tearDown, but please merge it anyway. +1!!!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set status to `closed`|
|-|-|

(In [20408]) Merge amp-ssl-test-2670

Author: therve
Reviewer: radix
Fixes [#2670](https://github.com/twisted/twisted/issues/2670)

Fix a SSL AMP test, that appears to fail after a long time but was in fact
completely broken.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
